graph/encoding/graphql/decode.go

```diff
@@ -0,0 +1,157 @@
+// Copyright ©2017 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package graphql
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	"gonum.org/v1/gonum/graph"
+	"gonum.org/v1/gonum/graph/encoding"
+)
+
+// Unmarshal parses the the JSON-encoded data and stores the result in dst.
+// Node IDs are obtained from the JSON fields identified by the uid parameter.
+// UIDs obtained from the JSON encoding must map to unique node ID values
+// consistently across the JSON-encoded spanning tree.
+func Unmarshal(data []byte, uid string, dst encoding.Builder) error {
+	if uid == "" {
+		return errors.New("graphql: invalid UID field name")
+	}
+	var src json.RawMessage
+	err := json.Unmarshal(data, &src)
+	if err != nil {
+		return err
+	}
+	gen := generator{dst: dst, uidName: uid, nodes: make(map[string]graph.Node)}
+	return gen.walk(src, nil, "")
+}
+
+// StringIDSetter is a graph node that can set its ID based on the given uid string.
+type StringIDSetter interface {
+	SetIDFromString(uid string) error
+}
+
+// LabelSetter is a graph edge that can set its label.
+type LabelSetter interface {
+	SetLabel(string)
+}
+
+type generator struct {
+	dst encoding.Builder
+
+	// uidName is the name of the UID field in the source JSON.
+	uidName string
+
+	nodes map[string]graph.Node
```

Comment:
Add a comment stating the key of the map; e.g.

```go
// nodes maps from node value to graph node.
```

Is this comment correct? Are node values unique?